cpp/examples/tutorial_examples/file_access_example.cc

```diff
@@ -0,0 +1,174 @@
+  int16_t years_raw[5] = {1990, 2000, 1995, 2000, 1995};
+  ARROW_RETURN_NOT_OK(int16builder.AppendValues(years_raw, 5));
+  std::shared_ptr<arrow::Array> years;
+  ARROW_ASSIGN_OR_RAISE(years, int16builder.Finish());
+
+  //Get a vector of our Arrays
+  std::vector<std::shared_ptr<arrow::Array>> columns = {days, months, years};
+
+  //Make a schema to initialize the Table with
+  std::shared_ptr<arrow::Field> field_day, field_month, field_year;
+  std::shared_ptr<arrow::Schema> schema;
+
+  field_day = arrow::field("Day", arrow::int8());
+  field_month = arrow::field("Month", arrow::int8());
+  field_year = arrow::field("Year", arrow::int16());
+
+  schema = arrow::schema({field_day, field_month, field_year});
+  //With the schema and data, create a Table
+  std::shared_ptr<arrow::Table> table; 
+  table = arrow::Table::Make(schema, columns);
+
+  //Write out test files in IPC, CSV, and Parquet for the example to use.
+  std::shared_ptr<arrow::io::FileOutputStream> outfile;
+  ARROW_ASSIGN_OR_RAISE(outfile, arrow::io::FileOutputStream::Open("test_in.ipc"));
+  ARROW_ASSIGN_OR_RAISE(std::shared_ptr<arrow::ipc::RecordBatchWriter> ipc_writer,
+                        arrow::ipc::MakeFileWriter(outfile, schema));
+  ARROW_RETURN_NOT_OK(ipc_writer->WriteTable(*table));
+  ARROW_RETURN_NOT_OK(ipc_writer->Close());
+
+  ARROW_ASSIGN_OR_RAISE(outfile, arrow::io::FileOutputStream::Open("test_in.csv"));
+  ARROW_ASSIGN_OR_RAISE(auto csv_writer, arrow::csv::MakeCSVWriter(
+          outfile, table->schema()));
+  ARROW_RETURN_NOT_OK(csv_writer->WriteTable(*table));
+  ARROW_RETURN_NOT_OK(csv_writer->Close());
+
+  ARROW_ASSIGN_OR_RAISE(outfile, arrow::io::FileOutputStream::Open("test_in.parquet"));
+  PARQUET_THROW_NOT_OK(parquet::arrow::WriteTable(*table,
+                                                  arrow::default_memory_pool(),
+                                                  outfile, 5));
+
+  return arrow::Status::OK();
+}
+
+arrow::Status RunMain() {
+  //Generate initial files for each format with a helper function -- don't worry,
+  //we'll also write a table in this example.
+  ARROW_RETURN_NOT_OK(GenInitialFile());
+ 
+  //Reading and writing from files
+
+  //First, we have to set up a ReadableFile object, which just lets us point our
+  //readers to the right data on disk. We'll be reusing this object, and rebinding
+  //it to multiple files throughout the example.
+  std::shared_ptr<arrow::io::ReadableFile> infile;
+  //Get "test_in.ipc" into our file pointer
+  ARROW_ASSIGN_OR_RAISE(infile,
+                          arrow::io::ReadableFile::Open("test_in.ipc",
+                                                        arrow::default_memory_pool()));
+  //Open up the file with the IPC features of the library, gives us a reader object.
+  ARROW_ASSIGN_OR_RAISE(auto ipc_reader,
+                        arrow::ipc::RecordBatchFileReader::Open(infile));
+  //Using the reader, we can read Record Batches. Note that this is specific to IPC;
```

Comment:
Created: https://issues.apache.org/jira/browse/ARROW-17401

Reply:
Resolving this and the following one, because it is very helpful information, but I want to keep things clear on what'll ultimately be a pretty long PR thread.